### PR TITLE
[TE][detection] Implement anomalies cache to speed up detection

### PR DESCRIPTION
**Pull request by akshayrai**
Running anomaly detection, especially on an entity alert, will require it to repeatedly fetch the complete anomaly hierarchy tree from the db over and over again for every sliding window. By caching these anomalies, we will significantly reduce the IO latency and speed up our detection. Moreover, this will also allow us to run dimension exploration for a larger set of combinations.

*PS: This PR only implements the in-memory cache*

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/algorithm/DimensionWrapper.java [hunk 1]
```diff
@@ -308,6 +314,33 @@ public DetectionPipelineResult run() throws Exception {
...
+  }
+
+  private void warmUpAnomaliesCache(List<MetricEntity> nestedMetrics) {
+    DatasetConfigDTO dataset = retrieveDatasetConfig(nestedMetrics);
+    long cachingPeriodLookback = config.getProperties().containsKey(PROP_CACHE_PERIOD_LOOKBACK) ?
+        MapUtils.getLong(config.getProperties(), PROP_CACHE_PERIOD_LOOKBACK) : ThirdEyeUtils
+        .getCachingPeriodLookback(dataset.bucketTimeGranularity());
+
+    long paddingBuffer = TimeUnit.DAYS.toMillis(1); // 100000
+    AnomalySlice anomalySlice = new AnomalySlice()
```
Comment (bxji) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/algorithm/DimensionWrapper.java]: Why don't we need to include filters and detection component names when initializing this anomaly slice?

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/algorithm/DimensionWrapper.java [hunk 2]
```diff
@@ -308,6 +314,33 @@ public DetectionPipelineResult run() throws Exception {
...
+        .withDetectionId(this.config.getId())
+        .withStart(startTime - cachingPeriodLookback - paddingBuffer)
+        .withEnd(endTime + paddingBuffer);
+    //.withFilters(this.metricEntity.getFilters())
+    //.withDetectionCompNames(Collections.singletonList(this.detectorName));
```
Comment (bxji) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/algorithm/DimensionWrapper.java]: Why don't we need to include filters and detection component names when initializing this anomaly slice?

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/algorithm/DimensionWrapper.java (hunk 1 above)
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/algorithm/DimensionWrapper.java]: Filter on non-child anomalies?

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/cache/builder/AnomaliesCacheBuilder.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
+package org.apache.pinot.thirdeye.detection.cache.builder;
```
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/cache/builder/AnomaliesCacheBuilder.java]: Add header.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/spi/model/AnomalySlice.java [hunk 4]
```diff
@@ -120,11 +119,46 @@ public AnomalySlice withIsTaggedAsChild(boolean isTaggedAsChild) {
...
+   *   c. current filters are empty or exactly match request filters.
+   *   d. current detectionComponentNames are empty or contains request detectionComponentNames
+   *
+   * Note: Empty filters will fetch all the anomalies regardless of the anomalous dimension.
+   * Similarly, empty detectionComponentNames will fetch all the anomalies.
+   */
+  public boolean containSlice(AnomalySlice requestSlice) {
+    if (requestSlice.getDetectionId() != this.detectionId
+        || requestSlice.getStart() < this.getStart()
+        || requestSlice.getEnd() > this.getEnd()) {
```
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/spi/model/AnomalySlice.java]: Just want to double check this is right inclusive.  Normally we use right exclusive.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/cache/builder/AnomaliesCacheBuilder.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+import org.apache.pinot.thirdeye.detection.spi.model.AnomalySlice;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ *  Essentially a fetcher for fetching anomalies from cache/datasource.
+ *  The cache holds anomalies information per Anomaly Slices
+ */
+public class AnomaliesCacheBuilder {
```
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/cache/builder/AnomaliesCacheBuilder.java]: Do you have some test results?
Normally MySQL should be pretty optimized.
What is the gain for building a cache layer?

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/algorithm/DimensionWrapper.java (hunk 2 above)
Comment (xiaohui-sun) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/algorithm/DimensionWrapper.java]: Filter on config Id is suffice.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/spi/model/AnomalySlice.java (hunk 4 above)
Comment (akshayrai) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/spi/model/AnomalySlice.java]: It shouldn't matter as we comparing slices.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/cache/builder/AnomaliesCacheBuilder.java (hunk 3 above)
Comment (akshayrai) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/cache/builder/AnomaliesCacheBuilder.java]: added

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/cache/builder/AnomaliesCacheBuilder.java (hunk 5 above)
Comment (akshayrai) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/cache/builder/AnomaliesCacheBuilder.java]: Sharing some details:

**Execution Time (Entity Alert):**

| Dimension Exploration (Top k) | Without Anomalies Cache | With Anomalies Cache |
| -- | -- |  -- |
|100 | 20 mins | 7.5 mins | 
|1000 | 6 hours | 1 hour |

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/algorithm/DimensionWrapper.java (hunk 1 above)
Comment (akshayrai) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/algorithm/DimensionWrapper.java]: By default, fetchAnomalies will only return parent anomalies.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/algorithm/DimensionWrapper.java (hunk 2 above)
Comment (akshayrai) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/algorithm/DimensionWrapper.java]: The idea is that running a specific detection will require all its past anomalies irrespective of filters or detection component names. So we fetch it once and cache it. The dimension exploration, for example, will then filter the relevant anomalies from this cache.

**Comment by akshayrai** [comment]
Rebasing to remove conflicts.

**Pull request merged**